### PR TITLE
[Fix] Fix knn multi-gpu bug

### PR DESCRIPTION
**Pull request by fangyixiao18**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Multi-gpu results and single gpu results
image
image

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmselfsup/engine/hooks/deepcluster_hook.py [hunk 1]
```diff
@@ -74,14 +74,15 @@ def deepcluster(self, runner) -> None:
...
-        features = self.extractor(runner.model.module)['feat']
+        features = self.extractor(runner.model.module)
         runner.model.train()
 
         # step 2: get labels
         if not self.dist_mode or (self.dist_mode and runner.rank == 0):
             clustering_algo = _clustering.__dict__[self.clustering_type](
                 **self.clustering_cfg)
             # Features are normalized during clustering
+            features = features['feat'].cpu().numpy()
```
Comment (YuanLiuuuuuu) [inline mmselfsup/engine/hooks/deepcluster_hook.py:85]: Clustering is applied on cpu?

File: tools/benchmarks/classification/knn_imagenet/test_knn.py [hunk 2]
```diff
@@ -122,16 +120,13 @@ def main():
         dataloader=dataset_cfg.val_dataloader, seed=args.seed)
 
     # build the model
+    # model is determined in this priority: checkpoint > init_cfg > random
     model = MODELS.build(cfg.model)
     model.init_weights()
 
-    # model is determined in this priority: init_cfg > checkpoint > random
     if hasattr(cfg.model.backbone, 'init_cfg'):
```
Comment (YuanLiuuuuuu) [inline tools/benchmarks/classification/knn_imagenet/test_knn.py]: This block can be deleted

File: tools/benchmarks/classification/knn_imagenet/test_knn.py [hunk 3]
```diff
@@ -122,16 +120,13 @@ def main():
...
         if getattr(cfg.model.backbone.init_cfg, 'type', None) == 'Pretrained':
-            logger.info(
-                f'Use pretrained model: '
-                f'{cfg.model.backbone.init_cfg.checkpoint} to extract features'
-            )
+            pass
     elif args.checkpoint is not None:
```
Comment (YuanLiuuuuuu) [inline tools/benchmarks/classification/knn_imagenet/test_knn.py]: Better assert ``args.checkpoint`` is not ``None``

File: mmselfsup/engine/hooks/deepcluster_hook.py (hunk 1 above)
Comment (fangyixiao18) [reply mmselfsup/engine/hooks/deepcluster_hook.py:85]: The faiss package need numpy input

File: tools/benchmarks/classification/knn_imagenet/test_knn.py (hunk 3 above)
Comment (fangyixiao18) [reply tools/benchmarks/classification/knn_imagenet/test_knn.py]: if we use init_cfg, the `args.checkpoint` can be `None`

**Review by YuanLiuuuuuu — APPROVED** [review]

**Pull request merged**